runtime/src/bank.rs

```diff
@@ -8512,40 +8580,38 @@ impl Bank {
             }
         }
 
-        self.feature_set = Arc::new(FeatureSet { active, inactive });
-        newly_activated
+        (FeatureSet { active, inactive }, newly_activated)
     }
 
     fn apply_builtin_program_feature_transitions(
         &mut self,
         only_apply_transitions_for_new_features: bool,
         new_feature_activations: &HashSet<Pubkey>,
     ) {
-        const FEATURES_AFFECTING_RBPF: &[Pubkey] = &[
-            feature_set::error_on_syscall_bpf_function_hash_collisions::id(),
-            feature_set::reject_callx_r10::id(),
-            feature_set::switch_to_new_elf_parser::id(),
-            feature_set::bpf_account_data_direct_mapping::id(),
-            feature_set::enable_alt_bn128_syscall::id(),
-        ];
-        if !only_apply_transitions_for_new_features
-            || FEATURES_AFFECTING_RBPF
-                .iter()
-                .any(|key| new_feature_activations.contains(key))
         {
+            let mut loaded_programs_cache = self.loaded_programs_cache.write().unwrap();
+            let upcoming_program_runtime_environment_v1 = loaded_programs_cache
+                .upcoming_program_runtime_environment_v1
+                .take()
+                .map(|(env, _programs_to_recompile)| env);
             let program_runtime_environment_v1 = create_program_runtime_environment(
                 &self.feature_set,
                 &self.runtime_config.compute_budget.unwrap_or_default(),
                 false, /* deployment */
                 false, /* debugging_features */
             )
             .unwrap();
-            let mut loaded_programs_cache = self.loaded_programs_cache.write().unwrap();
             if *loaded_programs_cache.program_runtime_environment_v1
                 != program_runtime_environment_v1
             {
-                loaded_programs_cache.program_runtime_environment_v1 =
-                    Arc::new(program_runtime_environment_v1);
+                loaded_programs_cache.program_runtime_environment_v1 = if let Some(upcoming) =
+                    upcoming_program_runtime_environment_v1
+                        .filter(|env| &program_runtime_environment_v1 == env.as_ref())
```

Comment:
The `upcoming_program_runtime_environment_v1` is `Some()` even if nothing has changed, then it simply is a copy of the current environment. And the `==` operator checks the content of each environment not the pointers.

Reply:
Since `program_runtime_environment_v1` is same as `upcoming`, wouldn't it be ok to assign `Arc::new(program_runtime_environment_v1)` to `loaded_programs_cache.program_runtime_environment_v1`? i.e. retain the original code instead of this condition block?

Reply:
Makes sense.